marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java

```diff
@@ -114,4 +113,68 @@ private String convertRowToJSONString(InternalRow record) {
         return jsonObjectWriter.toString();
     }
 
+    protected JsonNode determineIngestionEndpointParams(Map<String, String> params) {
+        ObjectMapper objectMapper = new ObjectMapper();
+
+        String ingestEndpointParams = params.containsKey("ingestendpointparams")?
+            params.get("ingestendpointparams"):objectMapper.createObjectNode().toString();
+
+        JsonNode endpointParamsJsonNode;
+        try {
+            endpointParamsJsonNode = objectMapper.readValue(ingestEndpointParams, JsonNode.class);
+        } catch (IOException e) {
+            throw new RuntimeException("Unable to parse ingestendpointparams, cause: " + e.getMessage(), e);
+        }
+
+        boolean doesNotHaveApiPath = !endpointParamsJsonNode.has("apiPath");
+        boolean hasWorkUnitOrEndpointState = endpointParamsJsonNode.has("workUnit") || endpointParamsJsonNode.has("endpointState");
+        if (doesNotHaveApiPath && hasWorkUnitOrEndpointState) {
+            throw new RuntimeException("Cannot override default endpointState and/or workUnit. Please provide custom endpoint.");
+        }
+
+        final String apiModulePath = (endpointParamsJsonNode.hasNonNull("apiPath") &&
+            endpointParamsJsonNode.get("apiPath").asText().length() > 0) ?
+            endpointParamsJsonNode.get("apiPath").asText() : "/data-hub/5/data-services/ingestion/bulkIngester.api";
+        logger.info("Will write to endpoint defined by: " + apiModulePath);
+        ((ObjectNode)endpointParamsJsonNode).put("apiPath", apiModulePath);
+
+        if (endpointParamsJsonNode.hasNonNull("endpointState") && endpointParamsJsonNode.get("endpointState").asText().length() > 0) {
+            ((ObjectNode)endpointParamsJsonNode).put("endpointState", endpointParamsJsonNode.get("endpointState").asText());
+        }
+        // TODO : remove the below else block after java-client-api 5.3 release
+        else {
+            ((ObjectNode)endpointParamsJsonNode).put("endpointState", "{}");
+        }
+
+        if (endpointParamsJsonNode.hasNonNull("workUnit") && endpointParamsJsonNode.get("workUnit").asText().length() > 0) {
+
+            JsonNode workUnitNode;
+            try {
+                workUnitNode = objectMapper.readValue(endpointParamsJsonNode.get("workUnit").asText(), JsonNode.class);
+            } catch (JsonProcessingException e) {
+                throw new RuntimeException("Unable to parse workUnit, cause: " + e.getMessage(), e);
+            }
+            ((ObjectNode) workUnitNode).put("uriprefix", uriPrefix);
+            ((ObjectNode)endpointParamsJsonNode).put("workUnit", workUnitNode.toString());
```

Comment:
Use set("workUnit", workUnitNode) instead. That returns the JSON structure instead of converting it into a string. You can then do setWorkUnit(new JacksonHandle(theNode)) and don't have to fiddle with converting things into strings. Same goes for endpointState.